hbase-server/src/main/java/org/apache/hadoop/hbase/quotas/RegionServerRpcQuotaManager.java

```diff
@@ -157,39 +157,44 @@ public OperationQuota getQuota(final UserGroupInformation ugi, final TableName t
   /**
    * Check the quota for the current (rpc-context) user. Returns the OperationQuota used to get the
    * available quota and to report the data/usage of the operation.
-   * @param region the region where the operation will be performed
-   * @param type   the operation type
+   * @param region   the region where the operation will be performed
+   * @param type     the operation type
+   * @param isAtomic whether the given operation is atomic
    * @return the OperationQuota
    * @throws RpcThrottlingException if the operation cannot be executed due to quota exceeded.
    */
-  public OperationQuota checkQuota(final Region region, final OperationQuota.OperationType type)
-    throws IOException, RpcThrottlingException {
+  public OperationQuota checkQuota(final Region region, final OperationQuota.OperationType type,
+    boolean isAtomic) throws IOException, RpcThrottlingException {
     switch (type) {
       case SCAN:
         return checkQuota(region, 0, 0, 1);
       case GET:
         return checkQuota(region, 0, 1, 0);
       case MUTATE:
-        return checkQuota(region, 1, 0, 0);
+        return checkQuota(region, 1, isAtomic ? 1 : 0, 0);
     }
     throw new RuntimeException("Invalid operation type: " + type);
   }
 
   /**
    * Check the quota for the current (rpc-context) user. Returns the OperationQuota used to get the
    * available quota and to report the data/usage of the operation.
-   * @param region  the region where the operation will be performed
-   * @param actions the "multi" actions to perform
+   * @param region   the region where the operation will be performed
+   * @param actions  the "multi" actions to perform
+   * @param isAtomic whether the given operation is atomic
    * @return the OperationQuota
    * @throws RpcThrottlingException if the operation cannot be executed due to quota exceeded.
    */
-  public OperationQuota checkQuota(final Region region, final List<ClientProtos.Action> actions)
-    throws IOException, RpcThrottlingException {
+  public OperationQuota checkQuota(final Region region, final List<ClientProtos.Action> actions,
+    boolean isAtomic) throws IOException, RpcThrottlingException {
     int numWrites = 0;
     int numReads = 0;
     for (final ClientProtos.Action action : actions) {
       if (action.hasMutation()) {
         numWrites++;
+        if (isAtomic) {
```

Comment:
i think you should also check the action MutateType here. We want to treat Increment and Append as reads as well, but those don't have to come from an atomic RegionAction